structuralcodes/materials/constitutive_laws.py

```diff
@@ -0,0 +1,165 @@
+        tol = 1.0e-6
+        if abs(eps) - self._eps_sy > tol:
+            return self._Eh
+        return self._E
+
+
+class ParabolaRectangle(ConstitutiveLaw):
+    """Class for parabola rectangle constitutive law"""
+
+    __materials__: t.Tuple[str] = ('concrete',)
+
+    def __init__(
+        self,
+        fc: float,
+        eps_0: float = -0.002,
+        eps_u: float = -0.0035,
+        n: float = 2.0,
+        name: t.Optional[str] = None,
+    ) -> None:
+        """Initialize an Elastic-Plastic Material"""
+        name = name if name is not None else "ParabolaRectangleLaw"
+        super().__init__(name=name)
+        self._fc = -abs(fc)
+        self._eps_0 = -abs(eps_0)
+        self._eps_u = -abs(eps_u)
+        self._n = n
+
+    def get_stress(self, eps: ArrayLike) -> ArrayLike:
+        """Return the stress given strain"""
+        eps = np.atleast_1d(np.asarray(eps))
+        # Parabolic branch
+        sig = self._fc * (1 - (1 - (eps / self._eps_0)) ** self._n)
+        # Rectangle branch
+        sig[eps < self._eps_0] = self._fc
+        # Zero elsewhere
+        sig[eps < self._eps_u] = 0
+        sig[eps > 0] = 0
+        return sig
+
+    def get_tangent(self, eps: ArrayLike) -> ArrayLike:
+        """Return the tangent given strain"""
+        eps = np.atleast_1d(np.asarray(eps))
+        # parabolic branch
+        tangent = (
+            self._n
+            * self._fc
+            / self._eps_0
+            * (1 - (eps / self._eps_0)) ** (self._n - 1)
+        )
+        # Elsewhere tangent is zero
+        tangent[eps < self._eps_0] = 0.0
+        tangent[eps > 0] = 0.0
+        return tangent
+
+
+class UserDefined(ConstitutiveLaw):
+    """Class for a user defined constitutive law
+    The curve is defined with positive and optionally negative
+    values. After the last value, the stress is mantained constant"""
+
+    __materials__: t.Tuple[str] = ('concrete', 'steel', 'rebars')
+
+    def __init__(
+        self, x: ArrayLike, y: ArrayLike, name: t.Optional[str] = None
+    ) -> None:
+        """Initialize a UserDefined constitutive law"""
+        name = name if name is not None else "UserDefinedLaw"
+        super().__init__(name=name)
+        x = np.atleast_1d(np.asarray(x))
+        y = np.atleast_1d(np.asarray(y))
+        if len(x) != len(y):
+            raise ValueError("The two arrays should have the same length")
+        if not np.any(x < 0):
+            # User provided only positive part, reflect in negative
+            self._x = np.concatenate((-np.flip(x), x))
+            self._y = np.concatenate((-np.flip(y), y))
+        else:
+            # User gave both positive and negative parts
+            self._x = x
+            self._y = y
+
+    def get_stress(self, eps: ArrayLike) -> ArrayLike:
+        """Return the stress given strain"""
+        eps = np.atleast_1d(np.asarray(eps))
+        sig = np.interp(eps, self._x, self._y, left=0, right=0)
```

Comment:
Here, `np.interp` returns `0` after the last values to the left and right, and not a constant value as described in the docstring. Perhaps to stick to contant value and slightly rewrite the call to `np.interp`?

Reply:
Discussed with @mortenengen: we would add the flag with following options:

1. drop to zero
2. constant
3. tangent
4. secant

Reply:
Implemented this feature.
It is still needed to discuss one topic: the ultimate strain (the strain that defines a ultimate strain condition for sectional analysis) should be defined by user? Or anyhow taken as the last value of x given as input?